scripts/syntax_gen.py

```diff
@@ -729,6 +721,126 @@ class SyntaxNode;
+    # Start the clone source file.
+    clonef = open(os.path.join(builddir, "AllSyntaxClone.cpp"), "w")
+    clonef.write(
+        """//------------------------------------------------------------------------------
+// AllSyntaxClone.cpp
+// All generated syntax node clone functionality
+//
+// SPDX-FileCopyrightText: Michael Popoloski
+// SPDX-License-Identifier: MIT
+//------------------------------------------------------------------------------
+#include "slang/syntax/AllSyntax.h"
+
+// This file contains all syntax generated clone implementation.
+// It is auto-generated by the syntax_gen.py script under the scripts/ directory.
+
+namespace slang::syntax::shallow {
+template<typename T>
+SyntaxNode* clone(const T& node, BumpAllocator& alloc) {
+    return alloc.emplace<T>(node);
+}
+}
+"""
+    )
+    clonef.write(
+        """namespace slang::syntax::deep {
+template<typename T>
+SyntaxNode* clone(const T& node, BumpAllocator& alloc) {
+    return alloc.emplace<T>(node);
+}
+SyntaxNode* clone(const SyntaxListBase&, BumpAllocator&) {
+    return nullptr;
+}
+"""
+    )
+    # Write out deepClone methods for each derived type.
+    for k, v in sorted(alltypes.items()):
+        if not v.final:
+            continue
+        if v.final:
+            clonef.write(
+                "SyntaxNode* clone(const {0}& node, BumpAllocator& alloc) {{\n".format(
+                    k
+                )
+            )
+            clonef.write("    return alloc.emplace<{0}>(\n".format(k))
+            if "kind" in v.argNames:
+                clonef.write("        node.kind,\n".format(m[1]))
+            for i, m in enumerate(v.combinedMembers):
+                if m[1] in v.notNullMembers:
+                    clonef.write(
+                        "        *slang::syntax::deepClone<{0}>(*node.{1}, alloc)".format(
+                            m[0][9:-2], m[1]
+                        )
+                    )
+                elif m[0].startswith("SyntaxList") or m[0].startswith(
+                    "SeparatedSyntaxList"
+                ):
+                    clonef.write(
+                        "        *slang::syntax::deepClone(node.{0}, alloc)".format(
+                            m[1]
+                        )
+                    )
+                else:
+                    clonef.write("        node.{0}".format(m[1]))
+                if i != len(v.combinedMembers) - 1:
+                    clonef.write(",\n")
+                else:
+                    clonef.write("\n")
+            clonef.write("    );\n")
+            clonef.write("}\n")
+    clonef.write("}\n\n")
+    clonef.write(
+        """namespace slang::syntax {
+
+struct CloneVisitor {
+    template<typename U>
+    using clone_t = decltype(std::declval<U>().clone(std::declval<BumpAllocator&>()));
+
+    template<typename T>
+    SyntaxNode* visit(const T& node, BumpAllocator& alloc) {
+        if constexpr (is_detected_v<clone_t, T>) {
+            return node.clone(alloc);
```

Comment:
Which types still have a .clone() member in them? It seems like you removed them all so this branch would be unnecessary. Maybe just the SyntaxListBase types? Perhaps those should be moved out into standalone methods too for consistency.

Reply:
I try to implement on this way, but it fails.
Due to the SyntaxNode is cast into SyntaxListBase and it lost type information on runtime.
Maybe std::variant can solve the issue, but the code change is much bigger.

Reply:
Why doesn't deepClone have the same problem?